core/src/connection/manager.rs

```diff
@@ -0,0 +1,479 @@
+    pub fn add<M>(&mut self, conn: Connection<M, H::Handler>, info: Connected<C>) -> ConnectionId
+    where
+        H: IntoConnectionHandler<C> + Send + 'static,
+        H::Handler: ConnectionHandler<
+            Substream = Substream<M>,
+            InEvent = I,
+            OutEvent = O,
+            Error = HE
+        > + Send + 'static,
+        <H::Handler as ConnectionHandler>::OutboundOpenInfo: Send + 'static,
+        TE: error::Error + Send + 'static,
+        HE: error::Error + Send + 'static,
+        I: Send + 'static,
+        O: Send + 'static,
+        M: StreamMuxer + Send + Sync + 'static,
+        M::OutboundSubstream: Send + 'static,
+        C: Send + 'static
+    {
+        let task_id = self.next_task_id;
+        self.next_task_id.0 += 1;
+
+        let (tx, rx) = mpsc::channel(4);
+        self.tasks.insert(task_id, TaskInfo {
+            sender: tx, state: TaskState::Established(info)
+        });
+
+        let task: Pin<Box<Task<Pin<Box<future::Pending<_>>>, _, _, _, _, _, _>>> =
+            Box::pin(Task::established(task_id, self.events_tx.clone(), rx, conn));
+
+        if let Some(executor) = &mut self.executor {
+            executor.exec(task);
+        } else {
+            self.local_spawns.push(task);
+        }
+
+        ConnectionId(task_id)
+    }
+
+    /// Notifies all managed connections of an event.
+    ///
+    /// This function is "atomic", in the sense that if `Poll::Pending` is
+    /// returned then no event has been sent.
+    #[must_use]
+    pub fn poll_broadcast(&mut self, event: &I, cx: &mut Context) -> Poll<()>
+    where
+        I: Clone
+    {
+        for task in self.tasks.values_mut() {
+            if let Poll::Pending = task.sender.poll_ready(cx) { // (*)
+                return Poll::Pending;
+            }
+        }
+
+        for task in self.tasks.values_mut() {
+            let cmd = task::Command::NotifyHandler(event.clone());
+            match task.sender.start_send(cmd) {
+                Ok(()) => {},
+                Err(ref err) if err.is_full() => unreachable!("by (*)"),
+                Err(_) => {},
+            }
+        }
+
+        Poll::Ready(())
+    }
+
+    /// Gets an entry for a managed connection, if it exists.
+    pub fn entry(&mut self, id: ConnectionId) -> Option<Entry<'_, I, C>> {
+        if let hash_map::Entry::Occupied(task) = self.tasks.entry(id.0) {
+            Some(Entry::new(task))
+        } else {
+            None
+        }
+    }
+
+    /// Gets information about a connection, if it is established.
+    pub fn connected(&self, id: &ConnectionId) -> Option<&Connected<C>> {
+        if let Some(task) = self.tasks.get(&id.0) {
+            match &task.state {
+                TaskState::Pending => None,
+                TaskState::Established(c) => Some(c),
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Provides an API similar to `Stream`, except that it cannot produce an error.
```

Comment:
[futures 0.3.1 `Stream`](https://docs.rs/futures/0.3.1/futures/stream/trait.Stream.html) is also not aware of the notion of an `Error`, no?

Reply:
I haven't reviewed the code yet, but this comment is probably a reliquate from the era of futures 0.1.

Reply:
Good point, thanks!